File: 프로그래머스_코틀린/level3/디스크 컨트롤러.kt
package 프로그래머스_코틀린.level3

class Solution {
    fun solution(jobs: Array<IntArray>): Int {
        var answer = 0
        var time = 0
        var readyQueue = jobs.toMutableList()
        while (readyQueue.size > 0){

            var currentJobList = readyQueue.filter { it ->
                it.first() <= time
            }.sortedBy{it.last()}

            if (currentJobList.isEmpty()){
                time ++
                continue
            } else {
                time += currentJobList.first()[1]
                answer += time - currentJobList.first()[0]
                readyQueue.remove(currentJobList.first())
            }
        }

        return answer / jobs.size
    }
}

File: 프로그래머스_파이썬/level3/디스크 컨트롤러.py
import heapq


def solution(jobs):
    answer, time = 0, 0
    total_jobs = len(jobs)
    while len(jobs) > 0:
        ready_queue = sorted([job for job in jobs if job[0] <= time], key=lambda x: x[1])
        if len(ready_queue) == 0:
            time += 1
            continue

        job = ready_queue[0]
        time += job[1]
        answer += time - job[0]
        jobs.remove(job)

    return answer // total_jobs